include/deal.II/base/mpi.templates.h

```diff
@@ -413,6 +414,92 @@ namespace Utilities
       internal::all_reduce(MPI_MIN, values, mpi_communicator, minima);
     }
 
+
+
+    template <typename T>
+    std::vector<T>
+    compute_set_union(const std::vector<T> &vec, const MPI_Comm &comm)
+    {
+#ifdef DEAL_II_WITH_MPI
+      // 1) collect vector entries and create union
+      std::vector<T> result = vec;
+
+      if (this_mpi_process(comm) == 0)
+        {
+          for (unsigned int i = 1; i < n_mpi_processes(comm); i++)
+            {
+              MPI_Status status;
+              const auto ierr_1 = MPI_Probe(MPI_ANY_SOURCE,
+                                            internal::Tags::compute_union,
+                                            comm,
+                                            &status);
+              AssertThrowMPI(ierr_1);
+
+              int        amount;
+              const auto ierr_2 =
+                MPI_Get_count(&status,
+                              internal::mpi_type_id(vec.data()),
+                              &amount);
+              AssertThrowMPI(ierr_2);
+
+              const unsigned int old_size = result.size();
+              result.resize(old_size + amount);
+
+              const auto ierr_3 = MPI_Recv(result.data() + old_size,
+                                           amount,
+                                           internal::mpi_type_id(vec.data()),
+                                           status.MPI_SOURCE,
+                                           status.MPI_TAG,
+                                           comm,
+                                           MPI_STATUS_IGNORE);
+              AssertThrowMPI(ierr_3);
+
+              std::sort(result.begin(), result.end());
+              result.erase(std::unique(result.begin(), result.end()),
+                           result.end());
+            }
+        }
+      else
+        {
+          const auto ierr = MPI_Send(vec.data(),
+                                     vec.size(),
+                                     internal::mpi_type_id(vec.data()),
+                                     0,
+                                     internal::Tags::compute_union,
+                                     comm);
+          AssertThrowMPI(ierr);
+        }
+
+      // 2) broadcast result
+      int size = result.size();
+      MPI_Bcast(&size, 1, MPI_INT, 0, comm);
+      result.resize(size);
+      MPI_Bcast(
+        result.data(), size, internal::mpi_type_id(vec.data()), 0, comm);
+
+      return result;
+#else
+      (void)comm;
+      return vec;
+#endif
+    }
+
+
+
+    template <typename T>
+    std::set<T>
+    compute_set_union(const std::set<T> &set_in, const MPI_Comm &comm)
+    {
+      // convert vector to set
+      std::vector<T> vector_in(set_in.begin(), set_in.end());
+
+      // perform operation to vector
+      const std::vector<T> vector_out = compute_set_union(vector_in, comm);
+
+      // convert vector to set
+      return std::set<T>(vector_out.begin(), vector_out.end());
```

Comment:
So you are paying the price of sorting and finding unique elements twice with this implementation. 
I would still prefer to have an `ArrayView` as input and `std::unordered_set` as output and let functions using it deal with the conversion from `std::unordered_set` to `std::vector` (or whatever the function needs to return). In particular, we expect the number of manifold ids and boundary ids to be small anyway such that the overhead doesn't really matter.

Reply:
> The problem with ArrayView and MPI datatypes is that we do not know the length of the result at the beginning.

The only thing I would change is to create a `std::set` from the intermediate `result` `std::vector` instead of sorting and finding unique elements in-place.
I won't request changes. If others prefer the current state, that's fine with me.

Reply:
It of course depends on the number of hits and the length of the list, but I would assume the vector variant with sorting to be faster than the set in case we have many insertions and few hits, and set maybe faster in case we have many hits and a lot of redundancy over MPI ranks. `set` is suprisingly slow compared to keeping a vector of integers sorted for small to medium sizes.

Reply:
I am really only arguing about the interface here and not about the implementation.

Reply:
I am pretty sure that this is the most efficient implementation. We've found in implementing `IndexSet` that the most efficient ways of finding a unique set of entries is to put them into a vector, sort, uniquefy, and then turn into a `std::set`.

For integers, there is really no particular reason to use `std::unordered_set` over `std::set`. The former uses a hash function -- but integers are fairly good hashes for themselves already.